drivers/i2c/i2c_esp32.c

```diff
@@ -430,19 +428,19 @@ static int IRAM_ATTR i2c_esp32_master_read(const struct device *dev, struct i2c_
 			if (ret < 0) {
 				return ret;
 			}
-			i2c_hal_read_rxfifo(&data->hal, read_pr, rd_filled);
-			msg->buf += rd_filled;
+			i2c_hal_read_rxfifo(&data->hal, msg_buf, rd_filled);
+			msg_buf += rd_filled;
+			msg_len -= rd_filled;
 		}
 
 		/* I2C master won't acknowledge the last byte read from the
 		 * slave device. Divide the read command in two segments as
 		 * recommended by the ESP32 Technical Reference Manual.
 		 */
-		if (msg->len == 1) {
+		if (msg_len == 1) {
 			cmd.ack_val = 1,
 			cmd.byte_num = 1,
-			msg->len = 0;
-			read_pr = msg->buf;
+			msg_len = 0;
 
 			i2c_hal_write_cmd_reg(&data->hal, cmd, data->cmd_idx++);
 			i2c_hal_write_cmd_reg(&data->hal, cmd_end, data->cmd_idx++);
```

Comment:
Can't help but notice this that block for the msg_len == 1 case is the exact same code as the block above, other than `cmd.ack_val = msg_len == 1;`  It's easy to merge the two blocks.  This also removes the need for the `if(rd_filled)`.

```c
        while (msg_len) {
                /* The smaller of the FIFO length or 1 less than the msg length */
                rd_filled = (msg_len > SOC_I2C_FIFO_LEN) ? SOC_I2C_FIFO_LEN : (msg_len - 1);

                /* I2C master won't acknowledge the last byte read from the
                 * slave device. Divide the read command in two segments as
                 * recommended by the ESP32 Technical Reference Manual.
                 */
                if (msg_len == 1) { // or rd_filled == 0
                        /* Final byte */
                        rd_filled = 1;
                        cmd.ack_val = 1;
                } else {
                        cmd.ack_val = 0;
                }
                cmd.byte_num = rd_filled;

                i2c_hal_write_cmd_reg(&data->hal, cmd, data->cmd_idx++);
                i2c_hal_write_cmd_reg(&data->hal, cmd_end, data->cmd_idx++);
                i2c_hal_enable_master_rx_it(&data->hal);
                ret = i2c_esp32_transmit(dev);
                if (ret < 0) {
                        return ret;
                }
                i2c_hal_read_rxfifo(&data->hal, msg_buf, rd_filled);
                msg_buf += rd_filled;
                msg_len -= rd_filled;
        }
```

Reply:
@xyzzy42 TBH, my goal in here was just to make the read/write bytes count to work. Logic stuff was left aside. Anyway, I'll update the logic as well. Thanks!